hilti/toolchain/include/ast/attribute.h

```diff
@@ -68,43 +114,103 @@ class Attribute : public Node {
     }
 
     /**
      * Factory for an attribute with no argument.
      *
-     * @param tag name of the attribute, including the leading `&`
+     * @param kind the attribute's internal representation
      * @param m meta data to associate with the node
      */
-    static auto create(ASTContext* ctx, const std::string& tag, const Meta& m = Meta()) {
-        return create(ctx, tag, nullptr, m);
-    }
+    static auto create(ASTContext* ctx, Kind kind, const Meta& m = Meta()) { return create(ctx, kind, nullptr, m); }
 
 protected:
-    Attribute(ASTContext* ctx, Nodes children, std::string tag, Meta m = Meta())
-        : Node(ctx, NodeTags, std::move(children), std::move(m)), _tag(std::move(tag)) {}
+    Attribute(ASTContext* ctx, Nodes children, Kind kind, Meta m = Meta())
+        : Node(ctx, NodeTags, std::move(children), std::move(m)), _kind(kind) {}
 
     std::string _dump() const override;
 
     HILTI_NODE_0(Attribute, final);
 
 private:
-    std::string _tag;
+    Kind _kind;
+    static inline std::map<std::string_view, Kind> _attr_map{
+        {"&eod", Attribute::Kind::EOD},
+        {"&until", Attribute::Kind::UNTIL},
+        {"&until-including", Attribute::Kind::UNTIL_INCLUDING},
+        {"&parse-at", Attribute::Kind::PARSE_AT},
+        {"&parse-from", Attribute::Kind::PARSE_FROM},
+        {"&size", Attribute::Kind::SIZE},
+        {"&max-size", Attribute::Kind::MAX_SIZE},
+        {"&ipv4", Attribute::Kind::IPV4},
+        {"&ipv6", Attribute::Kind::IPV6},
+        {"&type", Attribute::Kind::TYPE},
+        {"&count", Attribute::Kind::COUNT},
+        {"&synchronize", Attribute::Kind::SYNCHRONIZE},
+        {"&default", Attribute::Kind::DEFAULT},
+        {"&anonymous", Attribute::Kind::ANONYMOUS},
+        {"&internal", Attribute::Kind::INTERNAL},
+        {"&optional", Attribute::Kind::OPTIONAL},
+        {"&static", Attribute::Kind::STATIC},
+        {"&no-emit", Attribute::Kind::NO_EMIT},
+        {"&on-heap", Attribute::Kind::ON_HEAP},
+        {"&nosub", Attribute::Kind::NOSUB},
+        {"&cxxname", Attribute::Kind::CXXNAME},
+        {"&have_prototype", Attribute::Kind::HAVE_PROTOTYPE},
+        {"&priority", Attribute::Kind::PRIORITY},
+        {"&convert", Attribute::Kind::CONVERT},
+        {"&while", Attribute::Kind::WHILE},
+        {"&requires", Attribute::Kind::REQUIRES},
+        {"&byte-order", Attribute::Kind::BYTE_ORDER_},
+        {"&bit-order", Attribute::Kind::BIT_ORDER},
+        {"&chunked", Attribute::Kind::CHUNKED},
+        {"&originator", Attribute::Kind::ORIGINATOR},
+        {"&responder", Attribute::Kind::RESPONDER},
+        {"&try", Attribute::Kind::TRY},
+        {"&needed-by-feature", Attribute::Kind::NEEDED_BY_FEATURE},
+        {"&requires-type-feature", Attribute::Kind::REQUIRES_TYPE_FEATURE},
+        {"&always-emit", Attribute::Kind::ALWAYS_EMIT},
+        {"&transient", Attribute::Kind::TRANSIENT},
+        {"&anchor", Attribute::Kind::ANCHOR},
```

Comment:
As an aside, with this PR we can now see that `&anchor` does nothing, but its use doesn't even trigger a warning. I filed #1928 to clean this up.

Reply:
do you have reference on when this was valid or its utility removed? I didn't really understand what it was, it was one of those attributes that never got caught in tests except for using it internally a few times (I think like some of those `needed-by-feature` like ones, but those are more obvious)

Reply:
No, I do not recall what it was originally used for. For above issue I actually removed all mentions of `&anchor` from the code base, and the test suite still passes with that. ☹️

Nothing for you to do here, so resolving.